app/src/main/java/com/github/polybooks/core/databaseImpl/OLBookDatabase.kt

```diff
@@ -0,0 +1,230 @@
+class OLBookDatabase : BookDatabase {
+
+    override fun queryBooks(): BookQuery = OLBookQuery()
+
+    private inner class OLBookQuery : BookQuery {
+
+        private var ordering = DEFAULT
+
+        private var empty : Boolean = false
+        private var title : String? = null
+        private var isbn : String? = null
+
+        override fun onlyIncludeInterests(interests: Collection<Interest>): BookQuery {
+            System.err.println("Warning: onlyIncludeInterest not fully implemented for OLBookQuery")
+            this.empty = true
+            return this
+        }
+
+        override fun searchByTitle(title: String): BookQuery {
+            this.empty = false
+            this.title = title
+            this.isbn = null
+            return this
+        }
+
+        override fun searchByISBN13(isbn13: String): BookQuery {
+            this.empty = false
+            this.title = null
+            this.isbn = isbn13
+            return this
+        }
+
+        override fun withOrdering(ordering: BookOrdering): BookQuery {
+            this.ordering = ordering
+            return this
+        }
+
+        override fun getAll(): CompletableFuture<List<Book>> {
+            if (empty) return CompletableFuture.completedFuture(Collections.emptyList())
+            else if (title != null) {
+                return TODO("Search by title")
+            } else { assert(isbn != null)
+                val url = userISBN2URL(isbn!!)
+                if (url != null) {
+                    return url2json(url)
+                        .thenApplyAsync { listOf(parseBook(it)) }
+                        .exceptionally { exception ->
+                            if (exception is CompletionException && exception.cause is FileNotFoundException)
+                                return@exceptionally Collections.emptyList<Book>()
+                            else if (exception is CompletionException) throw exception.cause!!
+                            else throw exception
+                    }
+                } else {
+                    val future = CompletableFuture<List<Book>>()
+                    future.completeExceptionally(Exception("ISBN is not valid"))
+                    return future
+                }
+            }
+        }
+
+        override fun getN(n: Int, page: Int): CompletableFuture<List<Book>> {
+            if (n < 0 || page < 0) {
+                val future : CompletableFuture<List<Book>> = CompletableFuture()
+                future.completeExceptionally(
+                    IllegalArgumentException(
+                        if (n < 0) "Cannot return a negative ($n) number of results"
+                        else "Cannot return a negative ($page) page number"
+                    )
+                )
+                return future
+            }
+            return getAll()
+        }
+
+        override fun getCount(): CompletableFuture<Int> {
+            return getAll().thenApplyAsync { it.size }
+        }
```

Comment:
Same here for the apply async

Reply:
Moreover, I was doing the same thing before. Keep in mind that `getAll` will transform the data into our custom classes (which is useless computation in this case) and the count how many of them there are

Reply:
Fair point. I wasn't trying to do something optimised at the time.